bpf-gpl/xdp.c

```diff
@@ -65,11 +66,63 @@ int calico_xdp(struct xdp_md *xdp_ctx) {
 	}
 
 	// Parse packets and drop malformed and unsupported ones
-	if (parse_packet_ip(&ctx) == -2) {
-		return XDP_DROP;
+	switch (parse_packet_ip(&ctx)) {
+	case -1:
+		ctx.fwd.res = XDP_DROP;
+		goto deny;
+	case -2:
+		ctx.fwd.res = XDP_PASS;
+		goto allow;
 	}
 
+	tc_state_fill_from_iphdr(&ctx);
+
+	switch(tc_state_fill_from_nexthdr(&ctx)) {
+	case -1:
+		ctx.fwd.res = XDP_DROP;
+		goto deny;
+	case -2:
+		ctx.fwd.res = XDP_PASS;
+		goto allow;
+	}
+
+	// Allow a packet if it hits an entry in the failsafe map
+	if (is_failsafe_in(ctx.state->ip_proto, ctx.state->dport, ctx.state->ip_src)) {
+		CALI_DEBUG("Inbound failsafe port: %d. Skip policy\n", ctx.state->post_nat_dport);
+		ctx.state->pol_rc = CALI_POL_ALLOW;
+		goto allow;
+	}
+
+	// Jump to the policy program
+	CALI_DEBUG("About to jump to policy program.\n");
+	bpf_tail_call(xdp_ctx, &cali_jump, PROG_INDEX_POLICY);
+	if (CALI_F_HEP) {
```

Comment:
We only use this program for HEP ingress, so no need to check this.  Return from the tail call => no untracked policy => allow.